### PR TITLE
Agent Bootstrapping

### PR DESCRIPTION
**Pull request by realark**
Note: Merging against feature branch and not master.

This is phase 1 of the classloader changes. The goal of the PR is to allow setting up the java agent's classpath in premain. For now all agent classes still run in the system classloader. There are a few TODOs and FIXMEs in the diff which will be resolved before going into master.

Changes
* Reorganize Gradle subprojects
* Bootstrap agent by extracting embedded jars form `dd-java-agent.jar` in premain.

Gradle Project Changes
* `dd-java-agent:agent-bootstrap`: Runs on bootstrap loader. Contains api classes like opentracing, dd-trace-api
 `dd-java-agent:agent-tooling`: renaming of tooling. Contains dd-trace-ot and agent dependencies.
* `dd-java-agent:instrumentation`: A single project which depends on `agent-tooling` and all the instrumentation. Used to create a single shadowJar which contains both the core agent and the instrumentation.
* `dd-java-agent`: Bare-bones jar with no dependencies. Extracts the bootstrap and instrumentation jar, sets up the classpath and invokes agent and opentracing startup.

### CODE REVIEW

File: dd-java-agent-ittests/src/test/java/datadog/trace/agent/ShadowPackageRenamingTest.java [hunk 1]
```diff
@@ -24,10 +26,14 @@ public void agentDependenciesRenamed() throws Exception {
...
+        "AgentInstaller should reside in the tmp tooling jar: " + agentSource,
+        agentSource.matches(".*/agent-tooling-and-instrumentation[^/]*.jar"));
     Assert.assertEquals("DD guava dep must be loaded from agent jar.", agentSource, agentGuavaDep);
     Assert.assertNotEquals(
         "User guava dep must not be loaded from agent jar.", agentSource, userGuava);
   }
+
+  // TODO: Write test
+  // for every class in bootstrap jar:
+  //   assert class not present in agent jar
```
Comment (tylerbenson) [inline dd-java-agent-ittests/src/test/java/datadog/trace/agent/ShadowPackageRenamingTest.java]: To restate my comment from the last PR, this should just be a whitelist.

File: dd-java-agent/agent-tooling/agent-tooling.gradle [hunk 2]
```diff
@@ -1,9 +1,10 @@
 apply from: "${rootDir}/gradle/java.gradle"
 
 dependencies {
+  compile project(':dd-java-agent:agent-bootstrap')
+  compile project(':dd-trace-ot')
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/agent-tooling.gradle]: Maybe we want these to be `compileOnly`?  I believe that would prevent the need from excluding them and their dependencies everywhere else.

File: dd-java-agent/src/main/java/datadog/trace/agent/TracingAgent.java [hunk 3]
```diff
@@ -16,55 +16,146 @@
...
+      final JarFile bootStrapJar =
+          new JarFile(
+              extractToTmpFile(
+                  TracingAgent.class.getClassLoader(),
+                  "agent-bootstrap.jar.zip",
+                  "agent-bootstrap.jar"));
+
+      final ClassLoader agentClassLoader = TracingAgent.class.getClassLoader();
+
+      inst.appendToSystemClassLoaderSearch(bootStrapJar);
```
Comment (tylerbenson) [inline dd-java-agent/src/main/java/datadog/trace/agent/TracingAgent.java]: Do you want this to be boostrap now, or are you waiting for a later PR?

Comment (realark) [reply dd-java-agent/src/main/java/datadog/trace/agent/TracingAgent.java]: That's in the next one.

File: dd-java-agent/agent-tooling/agent-tooling.gradle (hunk 2 above)
Comment (realark) [reply dd-java-agent/agent-tooling/agent-tooling.gradle]: I'll look into that.

File: dd-java-agent-ittests/src/test/java/datadog/trace/agent/ShadowPackageRenamingTest.java (hunk 1 above)
Comment (realark) [reply dd-java-agent-ittests/src/test/java/datadog/trace/agent/ShadowPackageRenamingTest.java]: I'll follow up with you in the next PR.

**Pull request merged**